### ISSUE TITLE
OAP 0.3.0 - Spark 2.1.0 API related

### ISSUE THREAD

**Issue by Moonlit-Sailor** [Intel-bigdata/OAP#646]
### OAP 0.3.0 - Spark 2.1.0 API related (until commit 5b63c970bf8f6b)

1.  _org.apache.spark.sql.execution.datasources.FileFormat:_
**Before:**  
def inferSchema(sparkSession: SparkSession, options: Map[String, String],files: Seq[FileStatus]): Option[StructType]
**After:**     
def inferSchema: Option[StructType] 
**Reason:** 
Easy storage for data schema, we need to load schema from disk directly

2.  _org.apache.spark.sql.execution.datasources.OutputWriterFactory_
        _org.apache.spark.sql.execution.datasources.OutputWriter_
**Before:** 
**After:** 
OutputWriterFactory  def commitJob(taskResults: Array[WriteResult]): Unit
Add ‘writeInternal(row: InternalRow) and setPartitionString(str: String)’ method in OutputWriter
**Reason:**
OutputWriterFactory toggle the commitJob operation to write meta info back.
‘writeInternal()’ write the internal row to specific file format, ‘setPartitionString()’ enables oap writer set specific partition string.

3.	_org.apache.spark.sql.execution.datasources.FileSourceStrategy_
**Before:** 
**After:**
Add file format process in apply() method. 
**Reason:**
Make oap index compatible with parquet format, and initialize file format

4.	_org.apache.spark.sql.execution.DataSourceScanExec_
**Before:** 
**After:**
Modify ‘doProduceVectorized()’ method
**Reason:**
Filter unnecessary row id(s) in whole stage codegen for parquet vectorized reader.

5.	_org.apache.spark.sql.catalyst.expressions.UnsafeRow_
**Before:**
**After:**
Add UnsafeRow.setNumFields
**Reason:**
Read UnsafeRow from disk thread safely.

6.	_org.apache.spark.util.collection.BitSet_
**Before:** 
**After:**
Overload BitSet constructor
**Reason:**
Use an array to initialize BitSet field.

7.	_org.apache.spark.executor.Executor_
**Before:**
**After:** Add customInfoClassNameSeq, customManagerSeq.  Modify Heartbeat
**Reason:**
customManagerSeq handles corresponding heart beat message. Modify Heatbeat to add fiber info of each executor.

8.	_org.apache.spark.executor.custom.CustomManager_
**Before:**
**After:**
Add CustomManager trait
**Reason:**
Add ‘status(conf: SparkConf): String’ to get Fiber info from each executor

9.	_org.apache.spark.scheduler.SparkListenerBus_
**Before:** 
**After:** Modify SparkListenerBus, add ‘customInfoUpdateEvent’ and ‘oapIndexInfoUpdateEvent’ in ‘doPostEvent()’ method
**Reason:**
Deal with SparkListenerCustomInfoUpdate event and SparkListenerOapIndexInfoUpdate event.

10.	_org.apache.spark.scheduler.SparkListenerInterface_
**Before:**
**After:**
Add ‘onCustomInfoUpdate’ and ‘onOapIndexInfoUpdate’ method. Add class SparkListenerCustomInfoUpdate and SparkListenerOapIndexInfoUpdate
**Reason:**
SparkListenerCustomInfoUpdate  represents data fiber info of each executor, and SparkListenerOapIndexInfoUpdate represents index fiber info of each executor. ‘onCustomInfoUpdate’ deal with SparkListenerCustomInfoUpdate event, ‘onOapIndexInfoUpdate’ deal with SparkListenerOapIndexInfoUpdate event.

11.	_org.apache.spark.sql.execution.datasources.oap.listener_
**Before:**
**After:**
Add FiberInfoListener and OapIndexInfoListener to inherit from SparkListener
**Reason:**
FiberInfoListener and OapIndexInfoListener deal with the corresponding spark events.

12.	_org.apache.spark.Heartbeat_
**Before:**
**After:**
Add ‘customizedInfo’ field in HeartBeat class
**Reason:**
customizedInfo represents the fiber info of each executor.

13.	_org.apache.spark.HeartbeatReceiver_
**Before:**
**After:**
Deal with fiber info of heart beat message in ‘receiveAndReply()’ method
**Reason:**
Extract fiber info in heart beat and post it to listen bus.

14.	_org.apache.spark.scheduler.DAGScheduler_
**Before:**
**After:**
Add ‘getOapCacheLocs’method, modify ‘getPreferredLocsInternal’ to insert oap fiber cached locations. 
**Reason:**
‘getOapCacheLocs’method gets oap fiber cached locations

15.	_org.apache.spark.sql.SparkSqlAstBuilder_
**Before:**
**After:**
Add ‘visitOapCreateIndex’, ‘visitOapDropIndex’, ‘visitOap’, ‘visitIndexCols’, ‘visitIndexCol’, ‘visitIndexType’, ‘visitOapRefreshIndices’, ‘visitOapShowIndex’, ‘visitOapCheckIndex’, ‘visitOapEnableIndex’, ‘visitOapDisableIndex’ methods
**Reason:**
Add these method to enable index relative DML.

16.	_org.apache.spark.executor.Executor_
**Before:**
**After:**
Add toLongArray method
**Reason:**
Bloom filter needs to use the bit array to dump all bits to disk

17.	_sqlBase.g4_
**Before:**
**After:**
Add OAP index DML
**Reason:**
Enable create/drop/check/show/enable/disable oap index.

18.	_org.apache.spark.sql.execution.datasources.DataSource_
**Before:**
**After:**
Add OAP class name and short name in DataSource map and backward compatibility map.
**Reason:**
Add oap in data source.

19.	_org.apache.spark.sql.execution.datasources.DataSourceStrategy_
**Before:**
**After:**
Add ‘expressions.Like’ process in translateFilters() method
**Reason:**
Enable spark to push down STRING_LIKE operator.

**Comment by adrian-wang** [comment]
Can you split these into separate issues?

**Comment by adrian-wang** [comment]
No.16, Executor? I suppose it is BitSet?

**Comment by gczsjdy** [comment]
`CoarseGrainedExecutorBackend`? Although it will be removed in #623

**Comment by adrian-wang** [comment]
@gczsjdy it is your responsibility to close that corresponding issue when #623 is resolved. For now just comment status on that issue, when it is created.

**Comment by Moonlit-Sailor** [comment]
ok, I split these changes into separate issues. This issue could be closed.
